### DISCUSSION TITLE
Problems after upgrading wdio from v7 to v8

### DISCUSSION THREAD

**Discussion by NikolaSimsic** [webdriverio/webdriverio#10834] [Technical Discussion]
Hi, 

I encountered a bunch of problems after the upgrade, here are some:

1. So updated wdio to v8, here is how my package.json looks like:

v7:
```
        "@types/node-fetch": "^2.6.2",
        "@types/uuid": "^8.3.4",
        "@typescript-eslint/eslint-plugin": "^5.42.0",
        "@typescript-eslint/parser": "^5.42.0",
        "@wdio/appium-service": "^7.26.0",
        "@wdio/browserstack-service": "^7.27.0",
        "@wdio/cli": "^7.27.0",
        "@wdio/junit-reporter": "^7.26.0",
        "@wdio/local-runner": "^7.27.0",
        "@wdio/logger": "^7.26.0",
        "@wdio/mocha-framework": "^7.26.0",
        "@wdio/spec-reporter": "^7.26.0",
        "appium": "^1.22.3",
        "deepmerge": "^4.2.2",
        "eslint": "^8.26.0",
        "eslint-config-prettier": "^8.5.0",
        "eslint-plugin-import": "^2.26.0",
        "eslint-plugin-mocha": "^10.1.0",
        "eslint-plugin-promise": "^6.1.1",
        "eslint-plugin-wdio": "^7.25.3",
        "expect-webdriverio": "^3.5.3",
        "form-data": "^4.0.0",
        "node-fetch": "^2.6.7",
        "prettier": "^2.7.1",
        "ts-node": "^10.9.1",
        "typescript": "^4.8.4",
        "uuid": "^9.0.0",
        "wdio-native-app-compare-service": "^2.1.0",
        "wdio-wait-for": "^3.0.0"
```

Updated to v8:

```
"type": "module",
    "devDependencies": {
        "@types/node-fetch": "^2.6.4",
        "@types/uuid": "^9.0.2",
        "@typescript-eslint/eslint-plugin": "^6.2.0",
        "@typescript-eslint/parser": "^6.2.0",
        "@wdio/appium-service": "^8.13.13",
        "@wdio/browserstack-service": "^8.13.13",
        "@wdio/cli": "^8.13.14",
        "@wdio/junit-reporter": "^8.12.2",
        "@wdio/local-runner": "^8.13.13",
        "@wdio/logger": "^8.11.0",
        "@wdio/mocha-framework": "^8.13.13",
        "@wdio/spec-reporter": "^8.12.2",
        "appium": "^1.22.3",
        "deepmerge": "^4.3.1",
        "eslint": "^8.26.0",
        "eslint-config-prettier": "^8.5.0",
        "eslint-plugin-import": "^2.26.0",
        "eslint-plugin-mocha": "^10.1.0",
        "eslint-plugin-promise": "^6.1.1",
        "eslint-plugin-wdio": "^8.8.7",
        "expect-webdriverio": "^4.2.7",
        "form-data": "^4.0.0",
        "node-fetch": "^2.6.7",
        "prettier": "^2.7.1",
        "ts-node": "^10.9.1",
        "typescript": "^4.8.4",
        "uuid": "^9.0.0",
        "wdio-native-app-compare-service": "^2.1.0",
        "wdio-wait-for": "^3.0.6"

```

In one of my files I use to interact with browserstack that uses:

`import SevereServiceError from "webdriverio/build/utils/SevereServiceError.js";`

I am now getting error when try to run the tests:

```
ERROR @wdio/config:ConfigParser: Failed loading configuration file: file:///Users/tests.conf.ts: ERR_PACKAGE_PATH_NOT_EXPORTED  ./build/utils/SevereServiceError.js 
```

I can overcome this error by removing usage of SevereServiceError, but that is not great.

2. This issue encountered during test execution, could not find anywhere that these functions are being deprecated, what might cause these:

```
TypeError: browser.compareElement is not a function
TypeError: browser.compareScreen is not a function
```

Anyone has any idea what is going on with these two points?

Thanks!

**Comment by olivier-martin-sf** [comment]
Not a WDIO core team member but my guess is that you are trying to access an internal entry point that's not [exported anymore in v8](https://github.com/webdriverio/webdriverio/blob/e5977a11b1d5eb4e7a88bb198d683d9da369f8a4/packages/webdriverio/package.json#L50-L60). Most v8 packages uses [subpath export](https://nodejs.org/api/packages.html#subpath-exports) and you can only access what's listed in the package manifest for those packages. 

It seems that the `SevereServiceError` variable is [exported by the `index` file](https://github.com/webdriverio/webdriverio/blob/e5977a11b1d5eb4e7a88bb198d683d9da369f8a4/packages/webdriverio/src/index.ts#L25C14-L25C32) so just changing your import as `import SevereServiceError` from "webdriverio";` should fix the issue you are reporting

**Reply by NikolaSimsic** [reply]
Hi Olivier! 
Thanks a lot for reply, that indeed makes sense, I double checked and fixed the error! 
Now I encounter other error:
```
SevereServiceError in "bound onPrepare"
SevereServiceError: Something went wrong while uploading an app to BrowserStack: TypeError: FormData is not a constructor
```

**Comment by christian-bromann** [comment]
> import SevereServiceError from "webdriverio/build/utils/SevereServiceError.js";

Please import it via:

```ts
import { SevereServiceError } from 'webdriverio'
```

As mentioned here: https://webdriver.io/docs/customservices/#service-error-handling

> 2\. could not find anywhere that these functions are being deprecated

These commands might have been added by a service you are using.

**Reply by NikolaSimsic** [reply]
After updating import to following I managed to solve first error, but now I am facing suddenly:

```
SevereServiceError in "bound onPrepare"
SevereServiceError: Something went wrong while uploading an app to BrowserStack: TypeError: FormData is not a constructor
```

In this catch:

```
        if (e instanceof SevereServiceError) {
            throw new HookError(e.message, hook[0].name);
        }
```

**Reply by NikolaSimsic** [reply]
Found a solution,

Updated import, from:
import * as FormData from “form-data”;

to:
import { default as FormData } from "form-data";

**Reply by NikolaSimsic** [reply]
For this part:

```
TypeError: browser.compareScreen is not a function
TypeError: browser.compareElement is not a function
...
```

They come from wdio-native-app-compare-service which has not being updated, it was/remained latest available version.

This is how service is defined inside of a wdio config file:

```
    services: [
        [
            "native-app-compare",
            {
                autoSaveBaseline: true,
                baselineFolder: "images/expected",
                screenshotPath: "images",
                blockOutNavigationBar: true,
                blockOutIphoneHomeBar: true,
                blockOutStatusBar: true,
                imageNameFormat: "{tag}",
                savePerDevice: true,
            },
        ],
    ],
```


EDIT:

Solved, configuration was not merging services correctly from the other file.
